File: madcalc_desktop/lib/models/measurement_unit.dart
enum MeasurementUnit {
  centimeters,
  millimeters;

  String get label {
    switch (this) {
      case MeasurementUnit.centimeters:
        return 'cm';
      case MeasurementUnit.millimeters:
        return 'mm';
    }
  }

  double get millimetersPerUnit {
    switch (this) {
      case MeasurementUnit.centimeters:
        return 10;
      case MeasurementUnit.millimeters:
        return 1;
    }
  }

  int? parse(String rawValue) {
    final normalized = rawValue.trim().replaceAll(',', '.');
    if (normalized.isEmpty) {
      return null;
    }

    final value = double.tryParse(normalized);
    if (value == null) {
      return null;
    }

    return (value * millimetersPerUnit).round();
  }

  String format(int valueMm, {bool includeUnit = true}) {
    final String value;
    switch (this) {
      case MeasurementUnit.millimeters:
        value = '$valueMm';
      case MeasurementUnit.centimeters:
        final centimeters = valueMm / millimetersPerUnit;
        value = valueMm % 10 == 0
            ? centimeters.toStringAsFixed(0)
            : centimeters.toStringAsFixed(1).replaceAll('.', ',');
    }

    return includeUnit ? '$value $label' : value;
  }

  static MeasurementUnit fromRaw(String rawValue) {
    return values.firstWhere(
      (unit) => unit.name == rawValue,
      orElse: () => MeasurementUnit.centimeters,
    );
  }
}

File: MadCalc_iOS/Sources/Models/MeasurementUnit.swift
import Foundation

enum MeasurementUnit: String, CaseIterable, Codable, Identifiable {
    case centimeters
    case millimeters

    var id: String { rawValue }

    var label: String {
        switch self {
        case .centimeters:
            "cm"
        case .millimeters:
            "mm"
        }
    }

    var millimetersPerUnit: Double {
        switch self {
        case .centimeters:
            10
        case .millimeters:
            1
        }
    }

    func parse(_ rawValue: String) -> Int? {
        let normalized = rawValue
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: ",", with: ".")

        guard !normalized.isEmpty, let value = Double(normalized) else {
            return nil
        }

        return Int((value * millimetersPerUnit).rounded())
    }

    func format(_ valueMm: Int, includeUnit: Bool = true) -> String {
        let value: String

        switch self {
        case .millimeters:
            value = "\(valueMm)"
        case .centimeters:
            let centimeters = Double(valueMm) / millimetersPerUnit
            if valueMm.isMultiple(of: 10) {
                value = String(format: "%.0f", centimeters)
            } else {
                value = String(format: "%.1f", centimeters).replacingOccurrences(of: ".", with: ",")
            }
        }

        return includeUnit ? "\(value) \(label)" : value
    }
}
